spherical_splines/SHTOOLS/man/man1/makegrid2d.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MAKEGRID2D 1"
.TH MAKEGRID2D 1 "2009-08-18" "SHTOOLS 2.6" "SHTOOLS 2.6"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "MakeGrid2D"
.IX Header "MakeGrid2D"
.IP "MakeGrid2D \-" 14
.IX Item "MakeGrid2D -"
Create a 2D cylindrical map from a set of spherical harmonic coefficients.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 MakeGrid2D (" 24
.IX Item "SUBROUTINE MakeGrid2D ("
\&\s-1GRID\s0, \s-1CILM\s0, \s-1LMAX\s0, \s-1INTERVAL\s0, \s-1NLAT\s0, \s-1NLONG\s0, \s-1NORM\s0, \s-1CSPHASE\s0, F, A, \s-1NORTH\s0, \s-1SOUTH\s0, \s-1EAST\s0, \s-1WEST\s0 )
.RS 4
.IP "REAL*8" 19
.IX Item "REAL*8"
\&\s-1GRID\s0(180/INTERVAL+1, 360/INTERVAL+1), \s-1CILM\s0(2,\ \s-1LMAX+1\s0,\ \s-1LMAX+1\s0), \s-1INTERVAL\s0
.IP "\s-1INTEGER\s0" 19
.IX Item "INTEGER"
\&\s-1LMAX\s0, \s-1NLAT\s0, \s-1NLONG\s0
.IP "\s-1INTEGER\s0, \s-1OPTIONAL\s0" 19
.IX Item "INTEGER, OPTIONAL"
\&\s-1NORM\s0, \s-1CSPHASE\s0
.IP "REAL*8, \s-1OPTIONAL\s0" 19
.IX Item "REAL*8, OPTIONAL"
F, A, \s-1NORTH\s0, \s-1SOUTH\s0, \s-1EAST\s0, \s-1WEST\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIMakeGrid2D\fR will create a 2\-dimensional cylindrical map, equally spaced in (geocentric) latitude and longitude, from a set of input spherical harmonic coefficients. The output grid is in raster format possessing upper-left and lower-right coordinates of (90 N, 0 E) and (90 S, 360 E), repsectively. If the optional parameters \s-1NORTH\s0, \s-1SOUTH\s0, \s-1EAST\s0 and \s-1WEST\s0 are specified, then the output grid will possess upper-left and lower-right coordinates of (\s-1NORTH\s0, \s-1WEST\s0) and (\s-1SOUTH\s0, \s-1EAST\s0), repsectively. The employed spherical harmonic normalization and Condon-Shortley phase convention can be set by the optional arguments \s-1NORM\s0 and \s-1CSPHASE\s0; if not set, the default is to use geodesy 4\-pi normalized harmonics that exclude the Condon-Shortley phase of (\-1)^m.
.PP
If the optional arguments F and A are specified, the output function will be referenced to an ellipsoid with flattening F and semimajor axis A.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "\s-1GRID\s0" 11
.IX Item "GRID"
(output) REAL*8, \s-1DIMENSION\s0(180/INTERVAL+1, 360/INTERVAL+1)
.Sp
A 2D equally spaced map of the input spherical harmonic coefficients \s-1CILM\s0. The  array is in raster format with upper-left and lower-right coordinates of (90 N, 0 E) and (90 S, 360 E), respectively.
.IP "\s-1CILM\s0" 11
.IX Item "CILM"
(input) REAL*8, \s-1DIMENSION\s0 (2, \s-1LMAX+1\s0, \s-1LMAX+1\s0)
.Sp
The real spherical harmonic coefficients to be expanded in the space domain. The coefficients C1lm and C2lm refer to the \*(L"cosine\*(R" (Clm) and \*(L"sine\*(R" (Slm) coefficients, respectively, with Clm = \s-1CILM\s0(1, l+1, m+1) and Slm = \s-1CILM\s0(2, l+1, m+1). Alternatively, C1lm and C2lm correspond to the positive and negative order coefficients, respectively.
.IP "\s-1LMAX\s0" 11
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum spherical harmonic degree of the coefficients \s-1CILM\s0.
.IP "\s-1INTERVAL\s0" 11
.IX Item "INTERVAL"
(input) REAL*8
.Sp
The latitudinal and longitudinal spacing of \s-1GRID\s0.
.IP "\s-1NLAT\s0" 11
.IX Item "NLAT"
(output) \s-1INTEGER\s0
.Sp
The number of latitudinal samples. Both 90 N and 90 S are included.
.IP "\s-1NLONG\s0" 11
.IX Item "NLONG"
(output) \s-1INTEGER\s0
.Sp
The number of longitudinal samples. Both 0 and 360 E are included.
.IP "\s-1NORM\s0" 11
.IX Item "NORM"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
1 (default) = Geodesy 4\-pi normalized harmonics; 2 = Schmidt semi-normalized harmonics; 3 = unnormalized harmonics; 4 = orthonormal harmonics.
.IP "\s-1CSPHASE\s0" 11
.IX Item "CSPHASE"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
1 (default) = do not apply the Condon-Shortley phase factor to the associated Legendre functions; \-1 = append the Condon-Shortley phase factor of (\-1)^m to the associated Legendre functions.
.IP "F" 11
.IX Item "F"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The flattening of the reference ellipoid that is subtracted from the function. This is given by (R_equator \- R_pole) / R_equator. The semi-major axis A (i.e., R_equator) must be specified for this calculation.
.IP "A" 11
.IX Item "A"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The semi-major axis of the reference ellispoid that is subtracted from the function. The flattening F must be specified for this calculation.
.IP "\s-1NORTH\s0" 11
.IX Item "NORTH"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The maximum latitude of the output raster grid, in degrees. The default is 90 degrees.
.IP "\s-1SOUTH\s0" 11
.IX Item "SOUTH"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The minimum latitude of the output raster grid, in degrees. The default is \-90 degrees.
.IP "\s-1EAST\s0" 11
.IX Item "EAST"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The maximum longitude of the output raster grid, in degrees. The default is 360 degrees.
.IP "\s-1WEST\s0" 11
.IX Item "WEST"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The minimum longitude of the output raster grid, in degrees. The default is 0 degrees.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIshexpanddh\fR\|(1), \fImakegriddh\fR\|(1), \fIshexpandglq\fR\|(1), \fImakegridglq\fR\|(1), \fImakegravgrid2d\fR\|(1), \fImakemaggrid2d\fR\|(1)
.PP
<http://www.ipgp.jussieu.fr/~wieczor/SHTOOLS/SHTOOLS.html>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2005\-2007 by Mark Wieczorek <wieczor@ipgp.fr>.
.PP
This is free software; you can distribute and modify it under the terms of the revised \s-1BSD\s0 license.


spherical_splines/SHTOOLS/src/MakeGrid2D.f95
subroutine MakeGrid2D(grid, cilm, lmax, interval, nlat, nlong, norm, csphase, f, a, north, south, east, west)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	Given the Spherical Harmonic coefficients cilm, this subroutine
!	will compute a 2D grid with equal latitude and longitude spacings. 
!	Note that since this is NOT done using FFTs, this routine is therefore 
!	relatively SLOW! You might want to instead use MakeGridDH.
!
!	The value at grid(1,1) correspons to 90 degrees latitude
!	and 0 degrees longitude, and the longitude spacing goes from 0 to 360, 
!	with both points being calculated. If the optional
!	parameters NORTH, SOUTH, EAST and WEST are specified, the upper-left and lower right coordinates
!	of the output grid are (NORTH, WEST) and (SOUTH, EAST), respectively.
!
!	Calling Parameters:
!		IN
!			cilm:		Spherical harmonic coefficients.
!			lmax:		Maximum degree of expansions to be performed.
!			interval:	Spacing of output grid in DEGREES.
!		OUT
!			grid:		Gridded expansion of spherical harmonic coefficients.
!			nlat:		Number of latitude points for the grid.
!			nlong:		Number of longitude points for the grid.
!		OPTIONAL (IN)
!			norm		Spherical harmonic normalization:
!						(1) "geodesy" (default)
!						(2) Schmidt
!						(3) unnormalized
!						(4) orthonormalized
!			csphase:	1: Do not include the phase factor of (-1)^m
!					-1: Apply the phase factor of (-1)^m.
!			f, a:		Flattening and semimajor axis of the function. If included, 
!					an ellipsoid with these parameters will be subtracted from the
!					data.
!			north		Maximum latitude to compute, in degrees.
!			south		Minimum latitude to compute, in degrees.
!			east		Maximum longitude to compute, in degrees.
!			west		Minimum latitude to compute, in degrees.
!			
!
!	Notes:
!		1.	If lmax is greater than the the maximum spherical harmonic
!			degree of the input file, then this file will be ZERO PADDED!
!			(i.e., those degrees after lmax are assumed to be zero).
!
!	Dependencies:		PlmBar, PlBar, PlmSchmidt, PlSchmidt, PLegendreA, PLegendre, 
!				PlmON, CSPHASE_DEFAULT
!
!	Written by Mark Wieczorek (2003)
!
!	August 2005 - 	sines and cosines are now precomputed, leading to a savings in time
!			by a factor of up to 15 (if the code is compiled with normal
!			optimizations.
!	April 12, 2007	Added option to allow calculation of function referenced to a flattened ellipsoid.
!	July 21, 2007 	Reference ellipsoid is now calculated exactly using the flattening and semi-major
!			axis instead of just the second degree Legendre expansion coefficient.
!	August 19, 2007	Added option to calculate subdomains of a raster grid.
!
!	Copyright (c) 2005-2007, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	use SHTOOLS, only: PlmBar, PlBar, PlmSchmidt, PlSchmidt, PLegendreA, PLegendre, &
				PlmON, PlON, CSPHASE_DEFAULT

	implicit none

	real*8, intent(in) :: 	cilm(:,:,:), interval
	real*8, intent(out) :: 	grid(:,:)
	integer, intent(in) :: 	lmax
	integer, intent(out) :: nlat, nlong
	integer, intent(in), optional :: norm, csphase
	real*8, intent(in), optional :: f, a, north, south, east, west
	integer :: 		l, m, j, k, index, l1, m1, lmax_comp, phase, lnorm, temp, astat(4)
	real*8 :: 		pi, latmax, latmin, longmin, longmax, lat, longitude, &
				x, intervalrad,  r_ref
	real*8, allocatable ::	pl(:), cosm(:, :), sinm(:, :), cilm2(:,:, :)
	
	temp = 0
	if (present(north)) temp = temp + 1
	if (present(south)) temp = temp + 1
	if (present(east)) temp = temp + 1
	if (present(west)) temp = temp + 1
	
	if (temp /=0 .and. temp /=4) then
		print*, "Error --- MakeGrid2d"
		print*, "The optional parameters NORTH, SOUTH, EAST, and WEST must all be specified", &
			present(north), present(south), present(east), present(west)
		stop
	endif
	
	if (temp == 4) then
		latmax = north
		latmin = south
		longmin = west
		longmax = east
		
		if (latmax < latmin) then
			print*, "Error --- MakeGrid2d"
			print*, "NORTH must be larger than SOUTH."
			print*, "NORTH = ", latmax
			print*, "SOUTH = ", latmin
			stop
		endif
		
		if (latmax > 90.0d0 .or. latmin < -90.0d0) then
			print*, "Error --- MakeGrid2d"
			print*, "NORTH and SOUTH must lie between 90 and -90."
			print*, "NORTH = ", latmax
			print*, "SOUTH = ", latmin
			stop
		endif
		
		if (longmin > longmax) longmax = longmax + 360.0d0
		
	else
		latmax = 90.0d0
		latmin = -90.0d0
		longmin = 0.0d0
		longmax = 360.d0
	endif
	
	nlat = (latmax - latmin) / interval + 1
	nlong = (longmax - longmin) / interval + 1
	
	if (present(norm)) then
		if (norm > 4 .or. norm < 1) then
			print*, "Error - MakeGrid2d"
			print*, "Parameter NORM must be 1 (geodesy), 2 (Schmidt), 3 (unnormalized), or 4 (orthonormalized)."
			print*, "Input value is ", norm
			stop
		endif
		
		lnorm = norm
		
	else 
		lnorm = 1
	endif
	
 	if (present(csphase)) then
     		if (csphase /= -1 .and. csphase /= 1) then
     			print*, "MakeGrid2D --- Error"
     			print*, "CSPHASE must be 1 (exclude) or -1 (include)"
     			print*, "Input valuse is ", csphase
     			stop
     		else
     			phase = csphase
     		endif
     	else
     		phase = CSPHASE_DEFAULT
     	endif

    	
 	if (size(grid(:,1)) < nlat .or. size(grid(1,:)) < nlong ) then
		print*, "Error --- MakeGrid2D"
		print*, "GRID must be dimensioned ( (LATMAX-LATMIN)/INTERVAL+1, (LONGMAX-LONGMIN)/INTERVAL+1 ) where"
		print*, "INTERVAL = ", interval
		print*, "LATMAX = ", latmax
		print*, "LATMIN = ", latmin
		print*, "LONGMIN = ", longmin
		print*, "LONGMAX = ", longmax
		print*, "Input array is dimensioned ", size(grid(:,1)), size(grid(1,:))
		stop
	elseif (size(cilm(:,1,1)) < 2 .or. size(cilm(1,:,1)) < lmax+1 .or. size(cilm(1,1,:)) < lmax+1) then
		print*, "Error --- MakeGrid2D"
		print*, "CILM must be dimensioned as (2, LMAX+1, LMAX+1) where LMAX is ", lmax
		print*, "Input dimension is ", size(cilm(:,1,1)), size(cilm(1,:,1)), size(cilm(1,1,:))
		stop
	endif
	
	if ((present(f) .and. .not. present(a)) .or. (present(a) .and. .not. present(f)) ) then
		print*, "Error --- MakeGrid2D"
		print*, "Both F and A must be present"
		print*, "F ", present(f)
		print*, "A ", present(a)
		stop
	endif
	
	allocate(pl((lmax+1)*(lmax+2)/2), stat = astat(1))
	allocate(cosm(int(360./interval + 1), lmax+1), stat = astat(2))
	allocate(sinm(int(360./interval + 1), lmax+1), stat = astat(3))
	allocate(cilm2(2,lmax+1,lmax+1), stat = astat(4))
	if (astat(1) /= 0 .or. astat(2) /= 0 .or. astat(3) /=0 .or. astat(4) /= 0) then
		print*, "Error --- MakeGrid2D"
		print*, "Problem allocating arrays PL, COSM, SINM, and CILM2", astat(1), astat(2), astat(3), astat(4)
		stop
	endif
	
	pi = acos(-1.0d0)
	grid = 0.0d0
	
	lmax_comp = min(lmax, size(cilm(1,1,:))-1)

	intervalrad = interval*pi/180.0d0
	
	cilm2(1:2,1:lmax+1,1:lmax+1) = cilm(1:2,1:lmax+1,1:lmax+1)
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Precomputing sines and cosines leads to an increase in speed by a factor
	!	of almost 4 with no optimization, and by a factor of about 15 with normal 
	!	optimizations.
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

	do k=1, nlong
	
		longitude = longmin*pi/180.0d0 + dble(k-1)*intervalrad
		do m=0, lmax
			cosm(k, m+1) = cos(m*longitude)
			sinm(k, m+1) = sin(m*longitude)
		enddo
		
	enddo

	do j=1, nlat
	
		lat = latmax - dble(j-1)*interval
		x = sin(lat*pi/180.0d0)
		
		if (lat == 90.0d0 .or. lat == -90.0d0) then
		
			select case(lnorm)
				case(1); call PlBar(pl, lmax_comp, x)
				case(2); call PlSchmidt(pl, lmax_comp, x)
				case(3); call PLegendre(pl, lmax_comp, x)
				case(4); call PlON(pl, lmax_comp, x)
			end select
	
			do l = lmax_comp, 0, -1
				l1 = l+1
				grid(j,1) = grid(j,1) + cilm2(1,l1,1) * pl(l1)
			enddo
			
			grid(j, 2:nlong) = grid(j,1)
	
			if (present(f)) grid(j,1:nlong) = grid(j,1:nlong) - a*(1.0d0-f)
		
		else
		
			select case(lnorm)
				case(1); call PlmBar(pl, lmax_comp, x, csphase = phase)
				case(2); call PlmSchmidt(pl, lmax_comp, x, csphase = phase)
				case(3); call PLegendreA(pl, lmax_comp, x, csphase = phase)
				case(4); call PlmON(pl, lmax_comp, x, csphase = phase)
			end select
			
			do k = 1, nlong
			
				do l = lmax_comp, 0, -1
			
					l1 = l+1
					index = (l+1)*l/2 + 1
					grid(j,k) = grid(j,k) + cilm2(1,l1,1) * pl(index)
				
					do m = 1, l, 1
						m1 = m+1
						index = (l+1)*l/2 + m + 1
						grid(j,k) = grid(j,k) + ( cilm2(1,l1,m1)*cosm(k,m+1) + &
							cilm2(2,l1,m1)*sinm(k,m+1) ) * pl(index)
					enddo
				enddo
			
			enddo
		
			if (present(f)) then
				r_ref = a**2 * (1.0d0 + tan(lat*pi/180.0d0)**2) / &
					(1.0d0  + tan(lat*pi/180.0d0)**2 / (1.0d0 - f)**2 )
				r_ref = sqrt(r_ref)
				grid(j,1:nlong) = grid(j,1:nlong) - r_ref
			endif
			
		endif
		
	enddo
		
	
	! deallocate memory
	select case(lnorm)
		case(1); call PlmBar(pl, -1, x, csphase = phase)
		case(2); call PlmSchmidt(pl, -1, x, csphase = phase)
		case(4); call PlmON(pl, -1, x, csphase = phase)
	end select
	deallocate(pl)
	deallocate(cosm)
	deallocate(sinm)
	deallocate(cilm2)
	
end subroutine MakeGrid2D


